sdk/src/trace/batch_span_processor.cc

```diff
@@ -0,0 +1,184 @@
+
+    is_force_flush_ = true;
+    buffer_->Add(FORCE_FLUSH_TOKEN);
+
+    // wake up the worker thread
+    cv_.notify_one();
+
+    // wait on flush token span to be processed
+    std::unique_lock<std::mutex> force_flush_lk(force_flush_cv_m_);
+    auto status = force_flush_cv_.wait_for(force_flush_lk, timeout);
+
+    is_force_flush_ = false;
+
+    if(status == std::cv_status::timeout){
+        // TODO: glog timeout
+    } else {
+        // TODO: glog no timeout
+    }
+}
+
+void BatchSpanProcessor::DoBackgroundWork(){
+    int timeout = schedule_delay_millis_;
+
+    while (true)
+    {
+        // We wait for the buffer to fill up to max_export_batch_size 
+        //                     OR
+        // We see if the force_flush() method has been invoked
+        if(static_cast<int>(buffer_->size()) < max_export_batch_size_ && 
+           is_force_flush_ == false)
+           {
+            std::unique_lock<std::mutex> lk(cv_m_);
+            cv_.wait_for(lk, std::chrono::milliseconds(timeout));
+
+            // If shutdown has been invoked, break out of the loop
+            // and drain out the queue.
+            if(is_shutdown_) break;
+
+            // In case of spurious wake up, we export only if we have atleast one span in the batch.
+            // This is acceptable because batching is a best mechanism effort here.
+            if(static_cast<int>(buffer_->size()) == 0) continue;
+        }
+
+        auto start = std::chrono::steady_clock::now();
+        Export();
+        auto end = std::chrono::steady_clock::now(); 
+        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
+
+        timeout = schedule_delay_millis_ - duration.count();
+    }
+
+    // Drain the queue in case there are any spans still left. 
+    // Invoked when shutdown() is called.
+    DrainQueue();
+}
+
+void BatchSpanProcessor::Export(){
+    // Get the appropriate size
+    const int buffer_size = static_cast<int>(buffer_->size());
+    const int num_spans_to_export = buffer_size >= max_export_batch_size_ ? max_export_batch_size_ : buffer_size;
+
+    std::vector<std::unique_ptr<Recordable>> spans_arr;
+    
+    bool is_force_flush_token_present = false;
+
+    buffer_->Consume(
+    num_spans_to_export, [&](CircularBufferRange<AtomicUniquePtr<Recordable>> range) noexcept {
+        range.ForEach([&](AtomicUniquePtr<Recordable> &ptr) {
+        std::unique_ptr<Recordable> swap_ptr = std::unique_ptr<Recordable>(nullptr);
+        ptr.Swap(swap_ptr);
+
+        // Check if we just processed/consumed the force flush token
+        if(swap_ptr == nullptr) is_force_flush_token_present = true;
+
+        else spans_arr.push_back(std::unique_ptr<Recordable>(swap_ptr.release()));
+        return true;
+        });
+    });
+
+    exporter_->Export(nostd::span<std::unique_ptr<Recordable>>(spans_arr.data(), spans_arr.size()));
+
+    // If we processed the force flush token, notify
+    if (is_force_flush_token_present) 
```

Comment:
Why do we need force flush token? Can't we just use is_force_flush_ boolean? It looks like these two variables somewhat duplicate each other?

Reply:
Although I'll be getting rid of it entirely in my next commit, the reason why this was used is because the token acts as a marker for when the ForceFlush was called. If its processed, we know that we had flushed everything BEFORE the token, meaning the ForceFlush accomplished its purpose.

The boolean is_force_flush_ is only for the worker thread to know if the ForceFlush method was invoked. If it is true, it immediately calls Export().